### PR TITLE
rustdoc: avoid many `Symbol` to `String` conversions.

### PR DESCRIPTION
**Pull request by nnethercote**
Particularly when constructing file paths and fully qualified paths.
This avoids a lot of allocations, speeding things up on almost all
examples.

r? @GuillaumeGomez

### CODE REVIEW

**Comment by nnethercote** [comment]
In local measurements I see the following rustdoc improvements on rustc-perf:
- instructions: up to 5% better
- cycles: up to 8% better
- wall-time: up to 10% better

@bors try @rust-timer queue

**Comment by camelid** [comment]
I guess we had the same idea at the same time: #91876 :P

I'll close that one since the tests are failing and this covers it.

File: src/librustdoc/html/render/print_item.rs
Comment (camelid) [inline src/librustdoc/html/render/print_item.rs:44]: Please make this a doc comment.

File: src/librustdoc/html/tests.rs
Comment (camelid) [inline src/librustdoc/html/tests.rs:6]: typo:

```suggestion
    // the symbols used are static, and no `Symbol::intern` calls occur.
```

File: src/librustdoc/visit_ast.rs
Comment (camelid) [inline src/librustdoc/visit_ast.rs:49]: Ditto. Please use `get_opt_name` and remove `as_sym`.

File: src/librustdoc/html/format.rs
Comment (camelid) [reply src/librustdoc/html/format.rs:565]: It'd probably be helpful to look at my now-closed PR that used `UrlPartsBuilder`: #91876.

File: src/librustdoc/html/format.rs
Comment (nnethercote) [reply src/librustdoc/html/format.rs:522]: The extra `first` param just makes things more complicated. E.g. `syms` can be empty, but it can't be empty in `join_with_double_colon`. I agree it's ugly, but I wasn't able to see a way to make it simpler. I'd be happy to hear specific suggestions.

File: src/librustdoc/html/format.rs
Comment (nnethercote) [reply src/librustdoc/html/format.rs:565]: Is `UrlPartsBuilder` needed? `join_with_slash` does much the same thing, in much less code.

(Apologies again for the overlap in effort here.)

File: src/librustdoc/clean/inline.rs
Comment (nnethercote) [reply src/librustdoc/clean/inline.rs:169]: I just tried changing it to this:
```rust
    let s = elem.data.get_opt_name().unwrap();
    if !s.is_empty() { Some(s) } else { None }
```
And I get tons of test failures. Should I do it a different way?

Comment (nnethercote) [reply src/librustdoc/clean/inline.rs:169]: This works (passes tests), though I don't understand how it's equivalent to the `as_sym()` version:
```
    elem.data.get_opt_name().and_then(|s| if !s.is_empty() { Some(s) } else { None })

```

**Comment by nnethercote** [comment]
Instruction counts and cycles show clear wins, smaller than I got locally, probably because I was profiling without jemalloc enabled.  Wall times show no improvement in the default view, but if you show "insignificant" changes you can see up to 7%; I guess the wall times are so noisy in general that any non-huge change is considered insignificant.

**Comment by nnethercote** [comment]
I have pushed an update that addresses most of the comments.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
This is super cool, thanks! Let's wait for @camelid's confirmation as well.

**Comment by camelid** [comment]
I'll review this soon.

File: src/librustdoc/clean/inline.rs
Comment (camelid) [inline src/librustdoc/clean/inline.rs:169]: I don't think this `is_empty` check is necessary since it will be handled by `get_opt_name`.

File: src/librustdoc/html/format.rs
Comment (camelid) [reply src/librustdoc/html/format.rs:522]: Why do you need the extra `first` param? I find the interaction between this function and `href_with_root_path` somewhat confusing and complicated.

File: src/librustdoc/html/format.rs
Comment (camelid) [inline src/librustdoc/html/format.rs:528]: Won't this over-allocate in a bunch of situations? I imagine most paths are much less than 64 bytes.

File: src/librustdoc/html/format.rs
Comment (camelid) [inline src/librustdoc/html/format.rs:611]: Why not use `write!`? I think it'd be clearer.

File: src/librustdoc/html/format.rs
Comment (camelid) [inline src/librustdoc/html/format.rs:711]: I really think this code would be much clearer if it used `UrlPartsBuilder`.

File: src/librustdoc/visit_ast.rs
Comment (camelid) [inline src/librustdoc/visit_ast.rs:49]: Ditto.

**Review by camelid — CHANGES_REQUESTED** [review]
Overall, this looks good, but I have some code quality concerns.

File: src/librustdoc/clean/inline.rs
Comment (camelid) [reply src/librustdoc/clean/inline.rs:169]: cc #92032: I'm not sure what we should be doing here.

**Comment by camelid** [comment]
@nnethercote asked me on Zulip to take over this PR since they have other things to work on and I'm more familiar with rustdoc.

File: src/librustdoc/clean/inline.rs
Comment (camelid) [reply src/librustdoc/clean/inline.rs]: It's not equivalent to the `as_sym` version, but I think `as_sym`'s extra functionality is not needed.

File: src/librustdoc/clean/inline.rs
Comment (petrochenkov) [reply src/librustdoc/clean/inline.rs:169]: When both PRs are merged this will turn into just `elem.data.get_opt_name()`.

**Comment by camelid** [comment]
Rebased. Going to address the review comments a bit later.

File: src/librustdoc/html/url_parts_builder.rs
Comment (nnethercote) [inline src/librustdoc/html/url_parts_builder.rs:108]: Yes, jemalloc's size classes are as follows: 8, 16, 32, 48, 64, 80, 96, 12, 128, 160, 192, 224, 256, etc., so 8 is a good choice here. Might be worth adding something like "and jemalloc's size classes are all multiples of 8."  ?

Comment (camelid) [reply src/librustdoc/html/url_parts_builder.rs:108]: Sure, I can do that.

**Comment by camelid** [comment]
Perf is even better than before :)

**Comment by camelid** [comment]
I had to add some boxing (for iterators) to get the tests working; hopefully that won't affect perf.

@nnethercote does my jemalloc size class explanation look good?

File: src/librustdoc/html/url_parts_builder.rs
Comment (nnethercote) [inline src/librustdoc/html/url_parts_builder.rs:128]: It's not quite right... I suggest replacing the last two bullet points with this:
```
/// * jemalloc's size classes are all multiples of eight, which means that the amount of memory
///   it allocates will often match the amount requested, avoiding wasted bytes.
```

Comment (camelid) [reply src/librustdoc/html/url_parts_builder.rs]: Done, thanks!

**Comment by camelid** [comment]
@GuillaumeGomez could you re-review this sometime soon? It seems to be very prone to merge conflicts.

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:872]: I think you can put it on one line now. ;)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I'll make another quick review once the merge conflict is solved. Can you ping me once you have rebased @nnethercote please?

**Comment by camelid** [comment]
@GuillaumeGomez what merge conflict? I don't see any. Also, I've taken over this PR at @nnethercote's request.

File: src/librustdoc/html/render/print_item.rs
Comment (camelid) [reply src/librustdoc/html/render/print_item.rs]: Done.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I was sure there was a merge conflict. Oh well... Thanks for doing it! r=me once pass.

**Comment by camelid** [comment]
I just realized that I forgot to delete an outdated comment from an earlier commit (the one about `64 bytes is covers 99.9%+ of cases`). Done.

@bors r=GuillaumeGomez rollup=never

**Pull request merged**
